localstack/services/opensearch/cluster.py

```diff
@@ -118,6 +121,104 @@ def build_cluster_run_command(cluster_bin: str, settings: CommandSettings) -> Li
     return [cluster_bin] + cmd_settings
 
 
+class CustomEndpoint:
+    """
+    Encapsulates a custom endpoint (combines CustomEndpoint and CustomEndpointEnabled within the DomainEndpointOptions
+    of the cluster, i.e. combines two fields from the AWS OpenSearch service model).
+    """
+
+    enabled: bool
+    endpoint: str
+
+    def __init__(self, enabled: bool, endpoint: str) -> None:
+        """
+        :param enabled: true if the custom endpoint is enabled (refers to DomainEndpointOptions#CustomEndpointEnabled)
+        :param endpoint: defines the endpoint (i.e. the URL - refers to DomainEndpointOptions#CustomEndpoint)
+        """
+        self.enabled = enabled
+        self.endpoint = endpoint
+
+        if self.endpoint:
+            self.url = urlparse(endpoint)
+        else:
+            self.url = None
+
+
+def register_cluster(
+    host: str, path: str, forward_url: str, custom_endpoint: CustomEndpoint
+) -> List[Rule]:
+    """
+    Registers routes for a cluster at the edge router.
+    Depending on which endpoint strategy is employed, different routes are registered.
+    This method is tightly coupled with build_cluster_endpoint(). It prepares the necessary URL data already according
+    to the endpoint strategies. It also has an overview of the different endpoint strategies and their url schemas.
+
+    :param host: hostname of the inbound address without protocol or port
+    :param path: path of the inbound address
+    :param forward_url: whole address for outgoing traffic (including the protocol)
+    :param custom_endpoint: Object that stores a custom address and if its enabled.
+            If a custom_endpoint is set AND enabled, the specified address takes precedence
+            over any strategy currently active, and overwrites any host/path combination.
+
+    :return: A list of generated router rules. Primarily a reference to them for cleanup on shutdown.
+    """
+    # custom backends overwrite the usual forward_url
+    forward_url = config.OPENSEARCH_CUSTOM_BACKEND or forward_url
+    endpoint = ProxyHandler(forward_url)
+    rules = []
+    strategy = config.OPENSEARCH_ENDPOINT_STRATEGY
+    # custom endpoints override any endpoint strategy
+    if custom_endpoint and custom_endpoint.enabled:
+        LOG.debug(f"Registering route from {host}{path} to {endpoint.proxy.forward_base_url}")
+        assert not (
+            host == config.LOCALSTACK_HOSTNAME and (not path or path == "/")
+        ), "trying to register an illegal catch all route"
+        rules.append(
+            ROUTER.add(
+                path=path,
+                endpoint=endpoint,
+                host=f'{host}<regex("(:.*)?"):port>',
+            )
+        )
+        rules.append(
+            ROUTER.add(
+                f"{path}/<path:path>",
+                endpoint=endpoint,
+                host=f'{host}<regex("(:.*)?"):port>',
+            )
+        )
+    elif strategy == "domain":
+
```

Comment:
(very little) nit: This newline could be removed.

Reply:
:+1: